### PR TITLE
OG-395 Relaycall max calldata size, OG-412 accounting calldata gas cost in maxPossibleGas

### PR DESCRIPTION
**Pull request by shahafn**

### CODE REVIEW

File: src/relayserver/RelayServer.ts
Comment (drortirosh) [inline src/relayserver/RelayServer.ts:156]: you repeat yourself. how about "checkRange(title,min,max,value)"

File: src/relayserver/RelayServer.ts
Comment (drortirosh) [inline src/relayserver/RelayServer.ts:724]: toLowerCase. (not sure why the "locale" version was used here)

File: test/PaymasterCommitment.test.ts
Comment (drortirosh) [inline test/PaymasterCommitment.test.ts:224]: does it mean that the relay request (without the huge approval ) is 500 bytes long?

File: src/relayserver/RelayServer.ts
Comment (shahafn) [reply src/relayserver/RelayServer.ts:156]: Huh? That's not part of this PR

File: src/relayserver/RelayServer.ts
Comment (shahafn) [reply src/relayserver/RelayServer.ts:724]: Not Part of this PR, not sure who put `Locale` there

File: contracts/BasePaymaster.sol
Comment (forshtat) [reply contracts/BasePaymaster.sol:37]: I suggest keeping the original name

File: contracts/BasePaymaster.sol
Comment (forshtat) [reply contracts/BasePaymaster.sol:35]: I was under impression we agreed the limit will be nominated in gas, not in bytes.
So this would be calldata cost limit, was that changed?

File: contracts/RelayHub.sol
Comment (forshtat) [inline contracts/RelayHub.sol:130]: If the next version of a relay hub uses x2 the `_dataGasCostPerByte`, this Paymaster will not be able to update to use it.

File: contracts/RelayHub.sol
Comment (forshtat) [inline contracts/RelayHub.sol:226]: This was supposed to be changed so that `RejectedByPreRelayed` is considered top priority rule.

File: test/PaymasterCommitment.test.ts
Comment (forshtat) [inline test/PaymasterCommitment.test.ts:212]: I do not see a corresponding change in the Relay Hub, indicating this test is not correct.

File: test/RelayHubGasCalculations.test.ts
Comment (forshtat) [inline test/RelayHubGasCalculations.test.ts:357]: empty before is useless

File: test/RelayHubGasCalculations.test.ts
Comment (forshtat) [inline test/RelayHubGasCalculations.test.ts:350]: giving string aliases to these enums does not seem useful

File: test/RelayHubGasCalculations.test.ts
Comment (forshtat) [inline test/RelayHubGasCalculations.test.ts:371]: I don't see what 'half the length' refers to here

File: test/RelayHubGasCalculations.test.ts
Comment (forshtat) [inline test/RelayHubGasCalculations.test.ts:420]: there is an `assert.closeTo` check. I think it will be better fit as if our value is waay to big it may also be a problem

File: test/RelayHubGasCalculations.test.ts
Comment (forshtat) [inline test/RelayHubGasCalculations.test.ts:443]: why does this tests appear as entirely new?

**Review by forshtat — CHANGES_REQUESTED** [review]

File: src/relayserver/RelayServer.ts
Comment (drortirosh) [reply src/relayserver/RelayServer.ts:156]: I see it part of the diff.
you should create a PR against release, and then it will contain only its own content, and not other PRs, otherwise, I can't tell.

File: test/PaymasterCommitment.test.ts
Comment (drortirosh) [inline test/PaymasterCommitment.test.ts:213]: this is not the case Oleksii raised: preRelayedCall should succeed, and then forwarder should fail (on nonce error)
in this case, 22000 bytes of approvalData makes the paymaster pay (=relayCall emits "TransactionRelayed" instead of "RejectedByPaymaster"

File: src/relayserver/RelayServer.ts
Comment (shahafn) [reply src/relayserver/RelayServer.ts:156]: The only change is indentation.

File: contracts/BasePaymaster.sol
Comment (shahafn) [reply contracts/BasePaymaster.sol]: Reduced to 10K, please check with Augur etc. for a reasonable value

File: contracts/RelayHub.sol
Comment (shahafn) [reply contracts/RelayHub.sol]: Not in this PR

File: contracts/RelayHub.sol
Comment (shahafn) [reply contracts/RelayHub.sol:134]: This only validates paymasters' own requirement of data length, how does it change through hubs?

File: test/PaymasterCommitment.test.ts
Comment (shahafn) [reply test/PaymasterCommitment.test.ts:212]: See line 136 in RelayHub.sol. We added max calldata size, hub as this `require`. This is literally the edge case You asked me to add, when wee send the maximum size and paymaster rejects.

File: contracts/RelayHub.sol
Comment (drortirosh) [inline contracts/RelayHub.sol:136]: the unexpected value is the acceptanceBudget+dataCost, not the maxRelayExposure

File: contracts/interfaces/IPaymaster.sol
Comment (drortirosh) [inline contracts/interfaces/IPaymaster.sol:86]: .. and actual data size.

File: src/relayserver/ServerConfigParams.ts
Comment (drortirosh) [inline src/relayserver/ServerConfigParams.ts:85]: why? the whole idea is have a number here, e.g. 300k

Our tests then backtrack and make sure this number is viable.

File: contracts/BasePaymaster.sol
Comment (forshtat) [inline contracts/BasePaymaster.sol:35]: Why 10'0**2**0? Why not 10'000?

File: contracts/RelayHub.sol
Comment (forshtat) [inline contracts/RelayHub.sol:120]: Does it have to be part of an interface? I think this method can be internal.

File: src/relayserver/RelayServer.ts
Comment (forshtat) [inline src/relayserver/RelayServer.ts:210]: hey hey hey, this is a magic number! please use server parameters to define something like that.
as an option, can ask for block limit on a latest block when initializing

File: src/relayserver/RelayServer.ts
Comment (forshtat) [inline src/relayserver/RelayServer.ts:235]: also depends on `dataGasCost`, isn't it?

File: src/relayserver/ServerConfigParams.ts
Comment (forshtat) [inline src/relayserver/ServerConfigParams.ts:39]: This is coming from an unapproved branch, doesn't it? Where is the rest of 'blacklist' commit here?

File: test/PaymasterCommitment.test.ts
Comment (forshtat) [reply test/PaymasterCommitment.test.ts:212]: The description is incorrect then. This test "paymaster should not pay for requests with max msg.data size if it rejects in pre" is hard or impossible to implement. I cannot understand what is actually checked here.

File: test/PaymasterCommitment.test.ts
Comment (forshtat) [inline test/PaymasterCommitment.test.ts:239]: What is new in this test? How does it relate to the issue we are discussing?

File: test/PaymasterCommitment.test.ts
Comment (forshtat) [inline test/PaymasterCommitment.test.ts:292]: as mentioned earlier, this is not exposure that is high, that is call data size/ low Paymaster acceptanceBudget. Now this is an interesting test to add - raise this exception on call data size.

**Review by forshtat — CHANGES_REQUESTED** [review]

File: test/PaymasterCommitment.test.ts
Comment (drortirosh) [reply test/PaymasterCommitment.test.ts:224]: my calculations (for OG-416) show a minimum (dynamic buffers empty) relayCall is 22 words (4+ 22*32=708 bytes)
every non-empty string adds 1-word length and actual data ( **signature** is always non-empty, **req.data** is almost always non-empty (at least from solidity calls), **paymasterData**, **approvalData** are optional)

bottom line: I'm not sure you got to this "1000"

File: contracts/RelayHub.sol
Comment (shahafn) [reply contracts/RelayHub.sol:120]: Currently it's also being used in the server, and in general I believe that it might vary between hubs, and might be needed outside, whether on or off chain.

File: contracts/RelayHub.sol
Comment (shahafn) [reply contracts/RelayHub.sol:136]: fixed

File: src/relayserver/RelayServer.ts
Comment (shahafn) [reply src/relayserver/RelayServer.ts:210]: Right, though it's just a place holder for uint256 as parameter, since we only care about `msg.data.length` here.

File: src/relayserver/RelayServer.ts
Comment (shahafn) [reply src/relayserver/RelayServer.ts:232]: I didn't copy it. Just fixed an already existing check in the server. If we want to remove it, we should decide that outside of this PR.

File: src/relayserver/RelayServer.ts
Comment (shahafn) [reply src/relayserver/RelayServer.ts:235]: Right.

File: src/relayserver/ServerConfigParams.ts
Comment (shahafn) [reply src/relayserver/ServerConfigParams.ts:39]: Yes, it's part of another PR that this one is branched out from.

File: src/relayserver/ServerConfigParams.ts
Comment (shahafn) [reply src/relayserver/ServerConfigParams.ts:85]: I don't necessarily agree that it should be detached from how we got to that number. This default should be in line with the paymasters' and hubs' constants.

File: test/PaymasterCommitment.test.ts
Comment (shahafn) [reply test/PaymasterCommitment.test.ts:212]: You're the one who specifically asked me to add this test for the edge case where the msg.data.length reaches limit and paymaster rejects. Do you want me to remove it?

File: test/PaymasterCommitment.test.ts
Comment (shahafn) [reply test/PaymasterCommitment.test.ts:213]: See that below

File: test/PaymasterCommitment.test.ts
Comment (shahafn) [reply test/PaymasterCommitment.test.ts:239]: This is literally the attack raised by the audit.

File: test/PaymasterCommitment.test.ts
Comment (shahafn) [reply test/PaymasterCommitment.test.ts:292]: fixed

File: test/RelayHubGasCalculations.test.ts
Comment (shahafn) [reply test/RelayHubGasCalculations.test.ts:357]: fixd

File: test/RelayHubGasCalculations.test.ts
Comment (shahafn) [reply test/RelayHubGasCalculations.test.ts:350]: fixed

File: test/RelayHubGasCalculations.test.ts
Comment (shahafn) [reply test/RelayHubGasCalculations.test.ts:371]: Removed the comment anyway

File: test/RelayHubGasCalculations.test.ts
Comment (shahafn) [reply test/RelayHubGasCalculations.test.ts:420]: Fixed

File: test/RelayHubGasCalculations.test.ts
Comment (shahafn) [reply test/RelayHubGasCalculations.test.ts:443]: It doesn't, it's indentation

File: test/PaymasterCommitment.test.ts
Comment (forshtat) [reply test/PaymasterCommitment.test.ts:212]: Oh, ok. I understand now. Test description was not clear, or maybe I was drunk, I don't know.

File: contracts/RelayHub.sol
Comment (forshtat) [reply contracts/RelayHub.sol:120]: accepted

File: test/PaymasterCommitment.test.ts
Comment (forshtat) [reply test/PaymasterCommitment.test.ts:239]: That is correct, I misread the sentence

File: src/relayserver/RelayServer.ts
Comment (forshtat) [reply src/relayserver/RelayServer.ts:232]: agreed

**Comment by drortirosh** [comment]
I don't see any place that attempt to calculate `dataGasCostPerByte` -  I'd expect a  test to show in the logs how close is the value in the environment to the expected value.
I see tests that fail on `callDataSizeLimit` - but none that are based on the data-size calculation

**Comment by shahafn** [comment]
> `dataGasCostPerByte`

Look closer then.

**Comment by forshtat** [comment]
> > I don't see any place that attempt to calculate `dataGasCostPerByte` - I'd expect a test to show in the logs how close is the value in the environment to the expected value.
> > I see tests that fail on `callDataSizeLimit` - but none that are based on the data-size calculation
> 
> Look closer then.

Screenshot 2021-02-11 at 04 13 09

File: contracts/BasePaymaster.sol
Comment (forshtat) [reply contracts/BasePaymaster.sol:35]: put 10'500 please

**Review by forshtat — DISMISSED** [review]

**Comment by forshtat** [comment]
@drortirosh please re-review as well

File: test/RelayHubGasCalculations.test.ts
Comment (drortirosh) [inline test/RelayHubGasCalculations.test.ts:421]: You calculate `diffGas / absoluteDataLength`
you need to keep previous `dataLength` too for a meaningful calculation.

Comment (forshtat) [reply test/RelayHubGasCalculations.test.ts:421]: @shahafn this sounds correct

File: contracts/RelayHub.sol
Comment (shahafn) [reply contracts/RelayHub.sol]: fixed

File: test/RelayHubGasCalculations.test.ts
Comment (forshtat) [reply test/RelayHubGasCalculations.test.ts:421]: UPD: `dataLength` is not `absoluteDataLength` as it is just the 'additional' bytes. As this formula is always comparing with 0th gas usage value, it is actually a `diff` in this case

**Review by forshtat — APPROVED** [review]

**Pull request merged**
